channels/pkg/channels/apply.go

```diff
@@ -17,53 +17,114 @@ limitations under the License.
-	"os"
-	"os/exec"
-	"path"
-	"strings"
 
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/restmapper"
 	"k8s.io/klog/v2"
+	"k8s.io/kops/pkg/kubemanifest"
 )
 
-// Apply calls kubectl apply to apply the manifest.
-// We will likely in future change this to create things directly (or more likely embed this logic into kubectl itself)
-func Apply(data []byte) error {
-	// We copy the manifest to a temp file because it is likely e.g. an s3 URL, which kubectl can't read
-	tmpDir, err := os.MkdirTemp("", "channel")
+type Applier struct {
+	Client     dynamic.Interface
+	RESTMapper *restmapper.DeferredDiscoveryRESTMapper
+}
+
+// Apply applies the manifest to the cluster.
+func (p *Applier) Apply(ctx context.Context, manifest []byte) error {
+	objects, err := kubemanifest.LoadObjectsFrom(manifest)
 	if err != nil {
-		return fmt.Errorf("error creating temp dir: %v", err)
+		return fmt.Errorf("failed to parse objects: %w", err)
 	}
 
-	defer func() {
-		if err := os.RemoveAll(tmpDir); err != nil {
-			klog.Warningf("error deleting temp dir %q: %v", tmpDir, err)
+	objectsByKind := make(map[schema.GroupKind][]*kubemanifest.Object)
+	for _, object := range objects {
+		gv, err := schema.ParseGroupVersion(object.APIVersion())
+		if err != nil || gv.Version == "" {
+			return fmt.Errorf("failed to parse apiVersion %q", object.APIVersion())
+		}
+		kind := object.Kind()
+		if kind == "" {
+			return fmt.Errorf("failed to find kind in object")
 		}
-	}()
 
-	localManifestFile := path.Join(tmpDir, "manifest.yaml")
-	if err := os.WriteFile(localManifestFile, data, 0o600); err != nil {
-		return fmt.Errorf("error writing temp file: %v", err)
+		gvk := gv.WithKind(kind)
+		gk := gvk.GroupKind()
+		objectsByKind[gk] = append(objectsByKind[gk], object)
 	}
 
-	_, err = execKubectl("replace", "-f", localManifestFile, "--force", "--field-manager=kops")
-	return err
+	for gk := range objectsByKind {
+		if err := p.applyObjectsOfKind(ctx, gk, objectsByKind[gk]); err != nil {
+			return fmt.Errorf("failed to apply objects of kind %s: %w", gk, err)
+		}
+	}
+	return nil
 }
 
-func execKubectl(args ...string) (string, error) {
-	kubectlPath := "kubectl" // Assume in PATH
-	cmd := exec.Command(kubectlPath, args...)
-	env := os.Environ()
-	cmd.Env = env
+func (p *Applier) applyObjectsOfKind(ctx context.Context, gk schema.GroupKind, expectedObjects []*kubemanifest.Object) error {
+	klog.V(2).Infof("applying objects of kind: %v", gk)
+
+	restMapping, err := p.RESTMapper.RESTMapping(gk)
+	if err != nil {
+		return fmt.Errorf("unable to find resource for %s: %w", gk, err)
+	}
+
+	gvr := restMapping.Resource
 
-	human := strings.Join(cmd.Args, " ")
-	klog.V(2).Infof("Running command: %s", human)
-	output, err := cmd.CombinedOutput()
+	baseResource := p.Client.Resource(gvr)
+
+	actualObjects, err := baseResource.List(ctx, v1.ListOptions{})
```

Comment:
If we have to read the existing objects, we might want to scope by namespace, just for memory etc in big clusters.  We may not have to do this at all with server-side-apply, TBD maybe